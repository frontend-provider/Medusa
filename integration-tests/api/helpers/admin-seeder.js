const Scrypt = require("scrypt-kdf")
const { User } = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const buf = await Scrypt.kdf("secret_password", { logN: 1, r: 1, p: 1 })
  const password_hash = buf.toString("base64")

  await manager.insert(User, {
    id: "admin_user",
    email: "admin@medusa.js",
    api_token: "test_token",
    role: "admin",
    password_hash,
    ...data,
  })
}
