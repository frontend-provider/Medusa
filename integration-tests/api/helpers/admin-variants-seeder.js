const {
  Region,
  Product,
  ProductVariant,
  PriceList,
  CustomerGroup,
  Customer,
  Image,
  ShippingProfile,
  ProductCollection,
  ProductOption,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  const yesterday = ((today) => new Date(today.setDate(today.getDate() - 1)))(
    new Date()
  )

  const tenDaysAgo = ((today) => new Date(today.setDate(today.getDate() - 10)))(
    new Date()
  )

  const defaultProfile = await manager.findOne(ShippingProfile, {
    where: { type: ShippingProfile.default },
  })

  const collection = manager.create(ProductCollection, {
    id: "test-collection",
    handle: "test-collection",
    title: "Test collection",
  })

  await manager.save(collection)

  await manager.insert(Region, {
    id: "reg-europe",
    name: "Test Region Europe",
    currency_code: "eur",
    tax_rate: 0,
  })

  await manager.insert(Region, {
    id: "reg-us",
    name: "Test Region US",
    currency_code: "usd",
    tax_rate: 0,
  })

  const customer = await manager.create(Customer, {
    id: "test-customer",
    email: "john@doe.com",
    first_name: "John",
    last_name: "Doe",
    password_hash:
      "c2NyeXB0AAEAAAABAAAAAVMdaddoGjwU1TafDLLlBKnOTQga7P2dbrfgf3fB+rCD/cJOMuGzAvRdKutbYkVpuJWTU39P7OpuWNkUVoEETOVLMJafbI8qs8Qx/7jMQXkN", // password matching "test"
    has_account: true,
  })

  const customerGroup = await manager.create(CustomerGroup, {
    id: "test-group",
    name: "test-group",
  })

  await manager.save(customerGroup)
  customer.groups = [customerGroup]
  await manager.save(customer)

  const priceListActive = await manager.create(PriceList, {
    id: "pl",
    name: "VIP sale",
    description: "All year sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  await manager.save(priceListActive)

  const priceListExpired = await manager.create(PriceList, {
    id: "pl_expired",
    name: "VIP summer sale",
    description: "Summer sale for VIP customers.",
    type: "sale",
    status: "active",
    starts_at: tenDaysAgo,
    ends_at: yesterday,
  })

  await manager.save(priceListExpired)

  const priceListWithCustomers = await manager.create(PriceList, {
    id: "pl_with_customers",
    name: "VIP winter sale",
    description: "Winter sale for VIP customers.",
    type: "sale",
    status: "active",
  })

  priceListWithCustomers.customer_groups = [customerGroup]

  await manager.save(priceListWithCustomers)

  const productMultiReg = manager.create(Product, {
    id: "test-product",
    handle: "test-product-reg",
    title: "Multi Reg Test product",
    profile_id: defaultProfile.id,
    description: "test-product-description",
    status: "published",
    collection_id: "test-collection",
  })

  const image = manager.create(Image, {
    id: "test-image",
    url: "test-image.png",
  })

  productMultiReg.images = [image]

  await manager.save(productMultiReg)

  await manager.save(ProductOption, {
    id: "test-option",
    title: "test-option",
    product_id: "test-product",
  })

  const variantMultiReg = await manager.create(ProductVariant, {
    id: "test-variant",
    inventory_quantity: 10,
    title: "Test variant",
    variant_rank: 0,
    sku: "test-sku",
    ean: "test-ean",
    upc: "test-upc",
    barcode: "test-barcode",
    product_id: "test-product",
    prices: [
      {
        id: "test-price-multi-usd",
        currency_code: "usd",
        type: "default",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-usd",
        currency_code: "usd",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-expired-multi-usd",
        currency_code: "usd",
        amount: 70,
        price_list_id: "pl_expired",
      },
      {
        id: "test-price-multi-eur",
        currency_code: "eur",
        amount: 100,
      },
      {
        id: "test-price-discount-multi-eur",
        currency_code: "eur",
        amount: 80,
        price_list_id: "pl",
      },
      {
        id: "test-price-discount-multi-eur-with-customer",
        currency_code: "eur",
        amount: 40,
        price_list_id: "pl_with_customers",
      },
      {
        id: "test-price-discount-expired-multi-eur",
        currency_code: "eur",
        amount: 70,
        price_list_id: "pl_expired",
      },
    ],
    options: [
      {
        id: "test-variant-option-reg",
        value: "Default variant",
        option_id: "test-option",
      },
    ],
  })

  await manager.save(variantMultiReg)
}
