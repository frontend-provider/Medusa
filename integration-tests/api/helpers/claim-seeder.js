const {
  ClaimOrder,
  Order,
  LineItem,
  Fulfillment,
  Return,
} = require("@medusajs/medusa")

module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  let orderWithClaim = manager.create(Order, {
    id: "order-with-claim",
    customer_id: "test-customer",
    email: "test@email.com",
    payment_status: "captured",
    fulfillment_status: "fulfilled",
    billing_address: {
      id: "test-billing-address",
      first_name: "lebron",
    },
    shipping_address: {
      id: "test-shipping-address",
      first_name: "lebron",
      country_code: "us",
    },
    region_id: "test-region",
    currency_code: "usd",
    tax_rate: 0,
    discounts: [],
    payments: [
      {
        id: "test-payment-for-claim-order",
        amount: 10000,
        currency_code: "usd",
        amount_refunded: 0,
        provider_id: "test-pay",
        data: {},
      },
    ],
    items: [],
    ...data,
  })

  await manager.save(orderWithClaim)

  const li = manager.create(LineItem, {
    id: "test-item-co-2",
    fulfilled_quantity: 1,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 1,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li)

  const li2 = manager.create(LineItem, {
    id: "test-item-co-3",
    fulfilled_quantity: 4,
    title: "Line Item",
    description: "Line Item Desc",
    thumbnail: "https://test.js/1234",
    unit_price: 8000,
    quantity: 4,
    variant_id: "test-variant",
    order_id: orderWithClaim.id,
  })

  await manager.save(li2)

  const claimWithFulfillment = manager.create(ClaimOrder, {
    id: "claim-w-f",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  const ful1 = manager.create(Fulfillment, {
    id: "fulfillment-co-1",
    data: {},
    provider_id: "test-ful",
  })

  const ful2 = manager.create(Fulfillment, {
    id: "fulfillment-co-2",
    data: {},
    provider_id: "test-ful",
  })

  claimWithFulfillment.fulfillments = [ful1, ful2]

  await manager.save(claimWithFulfillment)

  const claimWithReturn = manager.create(ClaimOrder, {
    id: "claim-w-r",
    type: "replace",
    payment_status: "na",
    fulfillment_status: "not_fulfilled",
    order_id: "order-with-claim",
    ...data,
  })

  await manager.save(claimWithReturn)

  await manager.insert(Return, {
    id: "return-id-2",
    claim_order_id: "claim-w-r",
    status: "requested",
    refund_amount: 0,
    data: {},
  })
}
