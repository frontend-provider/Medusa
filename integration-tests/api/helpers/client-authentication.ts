const AUTH_COOKIE = {}
export async function getClientAuthenticationCookie(
  api,
  email = null,
  password = null
) {
  const user = {
    email: email ?? "test@medusajs.com",
    password: password ?? "test",
  }

  if (AUTH_COOKIE[user.email]) {
    return AUTH_COOKIE[user.email]
  }

  const authResponse = await api.post("/store/auth", user)
  AUTH_COOKIE[user.email] = authResponse.headers["set-cookie"][0].split(";")

  return AUTH_COOKIE[user.email]
}
