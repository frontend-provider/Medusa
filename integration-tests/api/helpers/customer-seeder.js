const { Customer, Address } = require("@medusajs/medusa")

module.exports = async (connection, data = {}) => {
  const manager = connection.manager

  await manager.insert(Customer, {
    id: "test-customer-1",
    email: "test1@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-2",
    email: "test2@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-3",
    email: "test3@email.com",
  })

  await manager.insert(Customer, {
    id: "test-customer-has_account",
    email: "test4@email.com",
    has_account: true,
  })

  await manager.insert(Address, {
    id: "test-address",
    first_name: "Lebron",
    last_name: "James",
    customer_id: "test-customer-1",
  })
}
