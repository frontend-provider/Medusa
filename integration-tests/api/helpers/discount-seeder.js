const {
  ShippingProfile,
  Region,
  Discount,
  DiscountRule,
} = require("@medusajs/medusa")
module.exports = async (dataSource, data = {}) => {
  const manager = dataSource.manager

  await manager.insert(Region, {
    id: "test-region",
    name: "Test Region",
    currency_code: "usd",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })

  await manager.insert(Region, {
    id: "test-region-2",
    name: "Test Region 2",
    currency_code: "eur",
    tax_rate: 0,
    payment_providers: [
      {
        id: "test-pay",
        is_installed: true,
      },
    ],
  })
}
