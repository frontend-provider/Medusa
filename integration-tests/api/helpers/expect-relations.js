const { expect } = require("@jest/globals")

export function expectRelations(expected, actual){
    expected.forEach(a => expectRelations_(a, actual))
}

function expectRelations_(expected, actual){
    const items = expected.split(".");
    let data_ = actual;

    for(const i in items){
        data_ = data_[items[i]];
        
        if(data_ instanceof Array && data_.length > 0) {
            data_ = data_[0];
        } 
    
        expect(data_).toBeDefined();
    }
}
