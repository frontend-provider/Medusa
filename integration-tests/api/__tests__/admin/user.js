const jwt = require("jsonwebtoken")
const path = require("path")

const setupServer = require("../../../helpers/setup-server")
const { useApi } = require("../../../helpers/use-api")
const { initDb, useDb } = require("../../../helpers/use-db")

const userSeeder = require("../../helpers/user-seeder")
const adminSeeder = require("../../helpers/admin-seeder")
const {
  simpleAnalyticsConfigFactory,
} = require("../../factories/simple-analytics-config-factory")
const startServerWithEnvironment =
  require("../../../helpers/start-server-with-environment").default

jest.setTimeout(30000)

const adminReqConfig = {
  headers: {
    Authorization: "Bearer test_token",
  },
}

describe("/admin/users", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    dbConnection = await initDb({ cwd })
    medusaProcess = await setupServer({ cwd })
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()
    medusaProcess.kill()
  })

  describe("GET /admin/users", () => {
    beforeEach(async () => {
      await adminSeeder(dbConnection)
      await userSeeder(dbConnection)
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("returns user by id", async () => {
      const api = useApi()

      const response = await api.get("/admin/users/admin_user", adminReqConfig)

      expect(response.status).toEqual(200)
      expect(response.data.user).toMatchSnapshot({
        id: "admin_user",
        email: "admin@medusa.js",
        api_token: "test_token",
        role: "admin",
        created_at: expect.any(String),
        updated_at: expect.any(String),
      })
    })

    it("lists users", async () => {
      const api = useApi()

      const response = await api
        .get("/admin/users", adminReqConfig)
        .catch((err) => {
          console.log(err)
        })

      expect(response.status).toEqual(200)

      expect(response.data.users).toMatchSnapshot([
        {
          id: "admin_user",
          email: "admin@medusa.js",
          api_token: "test_token",
          role: "admin",
          created_at: expect.any(String),
          updated_at: expect.any(String),
        },
        {
          id: "member-user",
          role: "member",
          email: "member@test.com",
          first_name: "member",
          last_name: "user",
          created_at: expect.any(String),
          updated_at: expect.any(String),
        },
      ])
    })
  })

  describe("POST /admin/users", () => {
    let user
    beforeEach(async () => {
      const api = useApi()
      await adminSeeder(dbConnection)
      await userSeeder(dbConnection)

      const response = await api
        .post(
          "/admin/users",
          {
            email: "test@forgottenPassword.com",
            role: "member",
            password: "test123453",
          },
          adminReqConfig
        )
        .catch((err) => console.log(err))

      user = response.data.user
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("creates a user", async () => {
      const api = useApi()

      const payload = {
        email: "test@test123.com",
        role: "member",
        password: "test123",
      }

      const response = await api
        .post("/admin/users", payload, adminReqConfig)
        .catch((err) => console.log(err))

      expect(response.status).toEqual(200)
      expect(response.data.user).toMatchSnapshot({
        id: expect.stringMatching(/^usr_*/),
        created_at: expect.any(String),
        updated_at: expect.any(String),
        role: "member",
        email: "test@test123.com",
      })
    })

    it("updates a user", async () => {
      const api = useApi()

      const updateResponse = await api
        .post(
          "/admin/users/member-user",
          { first_name: "karl" },
          adminReqConfig
        )
        .catch((err) => console.log(err.response.data.message))

      expect(updateResponse.status).toEqual(200)
      expect(updateResponse.data.user).toMatchSnapshot({
        id: "member-user",
        created_at: expect.any(String),
        updated_at: expect.any(String),
        role: "member",
        email: "member@test.com",
        first_name: "karl",
        last_name: "user",
      })
    })

    describe("Password reset", () => {
      it("Doesn't fail to fetch user when resetting password for an unknown email (unauthorized endpoint)", async () => {
        const api = useApi()

        const resp = await api.post("/admin/users/password-token", {
          email: "test-doesnt-exist@test.com",
        })

        expect(resp.status).toEqual(204)
      })

      it("Doesn't fail when generating password reset token (unauthorized endpoint)", async () => {
        const api = useApi()

        const resp = await api
          .post("/admin/users/password-token", {
            email: user.email,
          })
          .catch((err) => {
            console.log(err)
          })

        expect(resp.data).toEqual("")
        expect(resp.status).toEqual(204)
      })

      it("Resets the password given a valid token (unauthorized endpoint)", async () => {
        const api = useApi()

        const expiry = Math.floor(Date.now() / 1000) + 60 * 15
        const dbUser = await dbConnection.manager.query(
          `SELECT * FROM public.user WHERE email = '${user.email}'`
        )

        const token_payload = {
          user_id: user.id,
          email: user.email,
          exp: expiry,
        }
        const token = jwt.sign(token_payload, dbUser[0].password_hash)

        const result = await api
          .post("/admin/users/reset-password", {
            token,
            email: "test@forgottenpassword.com",
            password: "new password",
          })
          .catch((err) => console.log(err))

        const loginResult = await api.post("admin/auth", {
          email: "test@forgottenpassword.com",
          password: "new password",
        })

        expect(result.status).toEqual(200)
        expect(result.data.user).toEqual(
          expect.objectContaining({
            email: user.email,
            role: user.role,
          })
        )
        expect(result.data.user.password_hash).toEqual(undefined)

        expect(loginResult.status).toEqual(200)
        expect(loginResult.data.user).toEqual(
          expect.objectContaining({
            email: user.email,
            role: user.role,
          })
        )
      })

      it("Resets the password given a valid token without including email(unauthorized endpoint)", async () => {
        const api = useApi()

        const expiry = Math.floor(Date.now() / 1000) + 60 * 15
        const dbUser = await dbConnection.manager.query(
          `SELECT * FROM public.user WHERE email = '${user.email}'`
        )

        const token_payload = {
          user_id: user.id,
          email: user.email,
          exp: expiry,
        }
        const token = jwt.sign(token_payload, dbUser[0].password_hash)

        const result = await api
          .post("/admin/users/reset-password", {
            token,
            password: "new password",
          })
          .catch((err) => console.log(err.response.data.message))

        const loginResult = await api.post("admin/auth", {
          email: user.email,
          password: "new password",
        })

        expect(result.status).toEqual(200)
        expect(result.data.user).toEqual(
          expect.objectContaining({
            email: user.email,
            role: user.role,
          })
        )
        expect(result.data.user.password_hash).toEqual(undefined)

        expect(loginResult.status).toEqual(200)
        expect(loginResult.data.user).toEqual(
          expect.objectContaining({
            email: user.email,
            role: user.role,
          })
        )
      })

      it("Fails to Reset the password given an invalid token (unauthorized endpoint)", async () => {
        expect.assertions(2)
        const api = useApi()

        const token = "test.test.test"

        await api
          .post("/admin/users/reset-password", {
            token,
            email: "test@forgottenpassword.com",
            password: "new password",
          })
          .catch((err) => {
            expect(err.response.status).toEqual(400)
            expect(err.response.data.message).toEqual("invalid token")
          })
      })
    })
  })

  describe("DELETE /admin/users", () => {
    beforeEach(async () => {
      await adminSeeder(dbConnection)
      await userSeeder(dbConnection)
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("Deletes a user", async () => {
      const api = useApi()

      const userId = "member-user"

      const usersBeforeDeleteResponse = await api.get(
        "/admin/users",
        adminReqConfig
      )

      const usersBeforeDelete = usersBeforeDeleteResponse.data.users

      const response = await api.delete(`/admin/users/${userId}`, {
        headers: { Authorization: "Bearer test_token" },
      })

      const usersAfterDeleteResponse = await api.get(
        "/admin/users",
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data).toEqual({
        id: userId,
        object: "user",
        deleted: true,
      })

      const usersAfterDelete = usersAfterDeleteResponse.data.users

      expect(usersAfterDelete.length).toEqual(usersBeforeDelete.length - 1)
      expect(usersBeforeDelete).toEqual(
        expect.arrayContaining([expect.objectContaining({ id: userId })])
      )

      expect(usersAfterDelete).toEqual(
        expect.not.arrayContaining([expect.objectContaining({ id: userId })])
      )
    })
  })
})

describe("[MEDUSA_FF_ANALYTICS] /admin/analytics-config", () => {
  let medusaProcess
  let dbConnection

  beforeAll(async () => {
    const cwd = path.resolve(path.join(__dirname, "..", ".."))
    const [process, connection] = await startServerWithEnvironment({
      cwd,
      env: { MEDUSA_FF_ANALYTICS: true },
    })
    dbConnection = connection
    medusaProcess = process
  })

  afterAll(async () => {
    const db = useDb()
    await db.shutdown()

    medusaProcess.kill()
  })

  describe("DELETE /admin/users", () => {
    beforeEach(async () => {
      await adminSeeder(dbConnection)
      await userSeeder(dbConnection)
    })

    afterEach(async () => {
      const db = useDb()
      await db.teardown()
    })

    it("Deletes a user and their analytics config", async () => {
      const api = useApi()

      const userId = "member-user"

      await simpleAnalyticsConfigFactory(dbConnection, {
        user_id: userId,
      })

      const response = await api.delete(
        `/admin/users/${userId}`,
        adminReqConfig
      )

      expect(response.status).toEqual(200)
      expect(response.data).toEqual({
        id: userId,
        object: "user",
        deleted: true,
      })

      const configs = await dbConnection.manager.query(
        `SELECT * FROM public.analytics_config WHERE user_id = '${userId}'`
      )

      expect(configs).toMatchSnapshot([
        {
          created_at: expect.any(Date),
          updated_at: expect.any(Date),
          deleted_at: expect.any(Date),
          id: expect.any(String),
          user_id: userId,
          opt_out: false,
          anonymize: false,
        },
      ])
    })
  })
})
