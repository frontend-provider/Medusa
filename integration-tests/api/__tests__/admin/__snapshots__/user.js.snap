// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/admin/users GET /admin/users lists users 1`] = `
Array [
  Object {
    "api_token": "test_token",
    "created_at": Any<String>,
    "deleted_at": null,
    "email": "admin@medusa.js",
    "first_name": null,
    "id": "admin_user",
    "last_name": null,
    "metadata": null,
    "role": "admin",
    "updated_at": Any<String>,
  },
  Object {
    "api_token": null,
    "created_at": Any<String>,
    "deleted_at": null,
    "email": "member@test.com",
    "first_name": "member",
    "id": "member-user",
    "last_name": "user",
    "metadata": null,
    "role": "member",
    "updated_at": Any<String>,
  },
]
`;

exports[`/admin/users GET /admin/users returns user by id 1`] = `
Object {
  "api_token": "test_token",
  "created_at": Any<String>,
  "deleted_at": null,
  "email": "admin@medusa.js",
  "first_name": null,
  "id": "admin_user",
  "last_name": null,
  "metadata": null,
  "role": "admin",
  "updated_at": Any<String>,
}
`;

exports[`/admin/users POST /admin/users creates a user 1`] = `
Object {
  "api_token": null,
  "created_at": Any<String>,
  "deleted_at": null,
  "email": "test@test123.com",
  "first_name": null,
  "id": StringMatching /\\^usr_\\*/,
  "last_name": null,
  "metadata": null,
  "role": "member",
  "updated_at": Any<String>,
}
`;

exports[`/admin/users POST /admin/users updates a user 1`] = `
Object {
  "api_token": null,
  "created_at": Any<String>,
  "deleted_at": null,
  "email": "member@test.com",
  "first_name": "karl",
  "id": "member-user",
  "last_name": "user",
  "metadata": null,
  "password_hash": null,
  "role": "member",
  "updated_at": Any<String>,
}
`;

exports[`[MEDUSA_FF_ANALYTICS] /admin/analytics-config DELETE /admin/users Deletes a user and their analytics config 1`] = `
Array [
  Object {
    "anonymize": false,
    "created_at": Any<Date>,
    "deleted_at": Any<Date>,
    "id": Any<String>,
    "opt_out": false,
    "updated_at": Any<Date>,
    "user_id": "member-user",
  },
]
`;
